### PR TITLE
Merge Ready : Backport develop win_pkg to 2016.11 with additional bug fixes

### PR DESCRIPTION
**Pull request by damon-atkins**
### What does this PR do?
Backport current develop version of win_pkg with additional bug fixes.

Taken the develop version of win_pkg.py and
- import sys to get INT max size for compare_version()
- enhanced compare_version to support versions of 'Not Found' and 'latest'
  which can occur in windows.
- comment out import salt.utils.platform which is not in 2016.11 salt.utils.platform.is_windows() renamed back to salt.utils.is_windows()
- added back the warning about win_repo_source_dir removal as this is
  required for 2016.11 version.

Added on top of the backport is
- Additional comments into the code
-- Indicate pkg_list can return a 'version number' and also 'Not Found'
   version number
-- _get_latest_pkg_version() can return 'version number' or 'latest'
-- latest_version() can return blank if the latest version is installed or no
   version at all is installed
- upgrade_available did not require both version() and latest_version() as
  latest_version() does everything required.
- list_upgrades() added comments to explain code, simplified code
  to latest_version() rather than calling both latest_version()
  and compare_versions()
- list_available() is not required to force a refresh by default
- version() added comments and questions as the return value
  is non standard
- genrepo() add questions around the open mode and
  get_repo_data open mode
- _repo_process_pkg_sls()
-- file param was renamed to filename as file variable name classes
   with file()
-- rename var version to version_str so as not to import
   version() function
-- remove a section of code which was never executed and never used
- install()
-- simplified an if statements
-- add more code comments

### What issues does this PR fix or reference?
#43417 

### Behaviour Changes
In 2017 and develop the documentation style is different to 2016.11

The following are samples of changes, note this is the current output style in 2017 & develop:
```
#### pkg.version gimp nmap
New Results DIFFERENT          Original Results
{                               {
    "local": {               |      "local": {}
        "nmap": "",          <
        "gimp": ""           <
    }                        <
}                               }
```
2016.11 seems to follow the output style of existing pkg.* modules for non-windows platforms.
2017.7 and develop seem to follow a different style. Given this is a backport of develop above shows output style differences between 2016.11 original and 2017.7/develop

Given trying to keep the pkg.* interface the same across platforms, I would have thought the output should also be the same except for edge cases (e.g. pkg.refresh_db) ?

### Tests written?
No new tests

### CODE REVIEW

File: salt/modules/win_pkg.py [hunk 1]
```diff
@@ -692,10 +791,15 @@ def genrepo(**kwargs):
                     ret,
                     successful_verbose
                     )
-    with salt.utils.fopen(repo_details.winrepo_file, 'w+b') as repo_cache:
-        repo_cache.write(msgpack.dumps(ret))
+    serial = salt.payload.Serial(__opts__)
+    # TODO: 2016.11 has PY2 mode as 'w+b' develop has 'w+' ? PY3 is 'wb+'
+    # also the reading of this is 'rb' in get_repo_data()
+    mode = 'w+' if six.PY2 else 'wb+'
+    # with salt.utils.files.fopen(repo_details.winrepo_file, mode) as repo_cache:
```
Comment (terminalmage) [inline salt/modules/win_pkg.py]: This doesn't need to be here, it'll be caught once this would merge forward to develop and it's a simple matter of adding the `.files` to the function call.

File: salt/modules/win_pkg.py [hunk 2]
```diff
@@ -1656,9 +1787,11 @@ def get_repo_data(saltenv='base'):
         log.trace('get_repo_data called reading from disk')
 
     try:
+        serial = salt.payload.Serial(__opts__)
+        # with salt.utils.files.fopen(repo_details.winrepo_file, 'rb') as repofile:
```
Comment (terminalmage) [inline salt/modules/win_pkg.py]: Same here

File: salt/modules/win_pkg.py [hunk 3]
```diff
@@ -1718,12 +1854,30 @@ def compare_versions(ver1='', oper='==', ver2=''):
...
+        ver1 = str(sys.maxsize)
+    if ver2 == 'latest':
+        ver2 = str(sys.maxsize)
+    # Support version being the special meaning of 'Not Found'
+    if ver1 == 'Not Found':
+        ver1 = '0.0.0.0.0'
+    if ver2 == 'Not Found':
+        ver2 = '0.0.0.0.0'
+
+    # return salt.utils.versions.compare(ver1, oper, ver2, ignore_epoch=True)
```
Comment (terminalmage) [inline salt/modules/win_pkg.py]: This also doesn't need to be here.

**Review by terminalmage — COMMENTED** [review]
I'd like to get some feedback on this from @twangboy as he deals with this code more often than I. Nothing jumps out at me as egregious, just some unnecessary commented function calls.

**Comment by damon-atkins** [comment]
```
#### pkg.version grepwin saltenv=does_not_exist
New Results DIFFERENT 2017.7/develop      Original Results of 2016.11
{                                         {
    "local": {                         |      "local": ""
        "grepwin": "not available"     <
    }                                  <
}                                         }
```

Post 2016.11 the style of the results has changed?  I need feed back on this.
In the above example the doco for pkg.version in general across all platforms says it would return an empty string  "Original Results". In 2017 windows only returns a dict with 'not available' 

I suspect windows pkg should be raising CommandExecutionError and setting error.append(error msg),  which are then handle by pkg state. Rather than in the above case setting a dict with an error message.

CommandExecutionError is to big of a change in how errors are handled, to change in the existing releases.

File: salt/modules/win_pkg.py [hunk 4]
```diff
@@ -1346,32 +1448,27 @@ def remove(name=None, pkgs=None, version=None, **kwargs):
...
-    :type pkgs: list or None
-
     .. versionadded:: 0.16.0
 
-    *Keyword Arguments (kwargs)*
-    :param str saltenv: Salt environment. Default ``base``
-    :param bool refresh: Refresh package metadata. Default ``False``
+    Args:
+        name (str): The name(s) of the package(s) to be uninstalled. Can be a
+            single package or a comma delimited list of packages, no spaces.
```
Comment (twangboy) [inline salt/modules/win_pkg.py]: Multiline paramater docs need to start on their own line or they are not displayed properly. Needs to be:
```
Args:
    name (str):
        The name(s) of the packages(s)...
```

File: salt/modules/win_pkg.py [hunk 5]
```diff
@@ -1586,29 +1718,26 @@ def purge(name=None, pkgs=None, version=None, **kwargs):
...
-        A list of packages to delete. Must be passed as a python list. The
-        ``name`` parameter will be ignored if this option is passed.
+        name (str): The name of the package to be deleted.
 
-    *Keyword Arguments (kwargs)*
-    :param str saltenv: Salt environment. Default ``base``
-    :param bool refresh: Refresh package metadata. Default ``False``
+        version (str): The version of the package to be deleted. If this option
+            is used in combination with the ``pkgs`` option below, then this
+            version will be applied to all targeted packages.
```
Comment (twangboy) [inline salt/modules/win_pkg.py]: The docs for `version (str):` needs to start on the next line, same as above...

File: salt/modules/win_pkg.py [hunk 6]
```diff
@@ -1586,29 +1718,26 @@ def purge(name=None, pkgs=None, version=None, **kwargs):
...
 
-    .. versionadded:: 0.16.0
+        pkgs (list): A list of packages to delete. Must be passed as a python
+            list. The ``name`` parameter will be ignored if this option is
+            passed.
```
Comment (twangboy) [inline salt/modules/win_pkg.py]: Same here

**Review by twangboy — CHANGES_REQUESTED** [review]
A few docs changes

**Comment by damon-atkins** [comment]
re-run lint

**Comment by damon-atkins** [comment]
Go Go Jenkins!

**Comment by damon-atkins** [comment]
Hi @twangboy the doco changes have been completed.

**Comment by damon-atkins** [comment]
Hi @terminalmage the extra comments have been removed.

**Review by twangboy — APPROVED** [review]

**Comment by damon-atkins** [comment]
Hi @cachedout seems to be ready for merge.

**Comment by damon-atkins** [comment]
Hi Guys seems to be ready for merge.

**Comment by damon-atkins** [comment]
@cachedout any further reviews required?

File: salt/modules/win_pkg.py [hunk 7]
```diff
@@ -692,10 +789,15 @@ def genrepo(**kwargs):
...
+    serial = salt.payload.Serial(__opts__)
+    # TODO: 2016.11 has PY2 mode as 'w+b' develop has 'w+' ? PY3 is 'wb+'
+    # also the reading of this is 'rb' in get_repo_data()
+    mode = 'w+' if six.PY2 else 'wb+'
+
+    with salt.utils.fopen(repo_details.winrepo_file, mode) as repo_cache:
+        repo_cache.write(serial.dumps(ret))
     # save reading it back again. ! this breaks due to utf8 issues
-    #__context__['winrepo.data'] = ret
+    # __context__['winrepo.data'] = ret
```
Comment (cachedout) [inline salt/modules/win_pkg.py]: Please just remove this if it's not being used.

**Review by cachedout — CHANGES_REQUESTED** [review]

File: salt/modules/win_pkg.py (hunk 7 above)
Comment (damon-atkins) [reply salt/modules/win_pkg.py]: Updated the comment.

**Comment by cachedout** [comment]
My primary concern here is the size of the changeset that's going into a stable branch. I can't find any reference to win_pkg in the test suite either so there isn't much protection against regressions, either.

**Comment by damon-atkins** [comment]
Hi @cachedout  
122 (30%) doc string 23 comments (6%)  adds out of  408  
68 (25%) doc string  2 comments removes out of 269
I suggest you compare/diff against 2017 should be a significant drop in differences.
But also needs testing by someone else.

**Comment by damon-atkins** [comment]
Hi @UtahDave  are you available to do some independent testing on this.

**Review by cachedout — APPROVED** [review]
